### PR TITLE
[UI] Progress indicator for pending operations (e.g. duplicate)

### PR DESCRIPTION
**Pull request by akhenry**
# Summary of changes:
* Copy has been rewritten. It has been optimized, and the number of requests required for copying an object tree is significantly reduced. Copy is now performed bottom-up rather than top-down. This necessitated a move away from the CreationService which performs persistence and composition automatically which is incompatible with a bottom-up approach. 
* Copying is now three-stage. First, the entire object tree is duplicated in memory in a blocking operation. Once completed, persistence is performed as a background operation. Finally, the new duplicated object tree is added to the composition of its new parent. This prevents interaction with the duplicate objects until the entire process has successfully completed.
* Copying uses notifications and dialogs to indicate progress. This resulted in some very minor modifications to the dialog and notifications frameworks to make them a little more user friendly.
* Although copying is performed serially from the bottom up, all objects are persisted in parallel. It is assumed the browser will perform request batching for very large numbers of persist requests.

### Notes
* I would like to refactor copy again once CreationService supports in-memory creation of objects. The refactoring should be relatively minor.
* Copying is very fast, and consequently in order to see the progress notifications testing should be performed with large object trees using elastic search, with the chrome developer tools used to throttle the connection.

### CODE REVIEW

**Comment by akhenry** [comment]
Author Checklist
# Author Checklisk
* Changes address original issue? __Y__
* Unit tests included and/or updated with changes? __Y__
* Command line build passes? __Y__
* Expect to pass code review? __Y__

File: bundles.json [hunk 1]
```diff
@@ -23,7 +23,7 @@
     "platform/features/events",
     "platform/forms",
     "platform/identity",
-    "platform/persistence/local",
+    "platform/persistence/elastic",
```
Comment (VWoeltjen) [inline bundles.json]: Don't believe we want to include this change - think local storage should still be default persistence store for open source

Comment (akhenry) [inline bundles.json]: It absolutely should be local, apologies, I'll revert.

File: platform/entanglement/src/actions/CopyAction.js [hunk 2]
```diff
@@ -34,16 +34,102 @@ define(
...
+             is shown non-invasive banner notifications at the bottom of the screen.
+             */
+            if (phase.toLowerCase() === 'preparing' && !this.dialog){
+                this.dialog = this.dialogService.showBlockingMessage({
+                    title: "Preparing to copy objects",
+                    unknownProgress: true,
+                    severity: "info"
+                });
+            } else if (phase.toLowerCase() === "copying") {
+                this.dialogService.dismiss();
```
Comment (VWoeltjen) [inline platform/entanglement/src/actions/CopyAction.js]: Hm, this makes me want to rethink the dialogService API a little bit. Couple of things:

* `dismiss` is a private method of `dialogService`, so we shouldn't be relying on it here unless we make it public.
* The fact that `dismiss` is not connected to the specific dialog we previously displayed seems like it creates the opportunity for unusual behavior. I'm picturing some other blocking dialog being implemented in the same way (show the dialog on one event, dismiss it on another); if this were triggered at the same time as this dialog, then it would be very easy to dismiss the wrong dialog.

My thinking is that `showBlockingMessage` should return something that allows the specific dialog that was just shown to be dismissed. Either an object with a `dismiss` method, or perhaps just [a promise that we could pass into a `dialogService.cancel(promise)` function](http://www.bennadel.com/blog/2731-canceling-a-promise-in-angularjs.htm) (this would allow the other `dialogService` methods to also be cancelled consistently)
  
This is a little bit out of scope for this issue, so we could defer this into a follow-up issue, but since this is the first release usage of notifications we could also justify refactoring here. I'm comfortable with either option

File: platform/entanglement/src/services/CopyService.js [hunk 3]
```diff
@@ -54,45 +57,156 @@ define(
...
+                        model: makeClone(originalObject.getModel()),
+                        persistenceSpace: originalParent.getCapability('persistence')
+                    };
+                delete modelClone.model.composition;
+                delete modelClone.model.persisted;
+                delete modelClone.model.modified;
+                return $q.when(originalObject.useCapability('composition')).then(function(composees){
+                    progress({phase: "preparing"});
+                    return (composees || []).reduce(function(promise, composee){
+                            //If the object is composed of other
```
Comment (VWoeltjen) [inline platform/entanglement/src/services/CopyService.js]: Think this function (and perhaps a couple of others in this area) should get hoisted out into named functions for readability; nesting here gets a little deep, also resulting in some long lines. I do think the comments provide a helpful narrative here, so I'd keep those.

File: platform/entanglement/src/services/CopyService.js [hunk 4]
```diff
@@ -54,45 +57,156 @@ define(
...
+                        clones.push(modelClone);
+                        return modelClone;
+                    });
+                });
             }
+            
+            return copy(domainObject, parent).then(function(domainObjectClone){
+                //If the domain object being cloned is not a link, set its
+                // location to the new parent
+                if ( !(domainObject.hasCapability("location") && domainObject.getCapability("location").isLink())) {
```
Comment (VWoeltjen) [inline platform/entanglement/src/services/CopyService.js]: If I'm reading correctly, I think this is a behavioral change - duplicating a link would result in a copy of the linked object, but with the same `location` property as the original. This creates an inconsistency in the data: Assume A contains and locates B, and B is linked to from C. C gets copied, becoming C' and containing B'. B' says it's located in A, but A doesn't contain B'.

I believe that the existing behavior would have B' located by C' in this case, so I think we should remain consistent with that (IIRC this is by design)

File: platform/entanglement/src/services/CopyService.js [hunk 5]
```diff
@@ -54,45 +57,156 @@ define(
...
+         * memory
+         * @param domainObject The original object to be copied
+         * @param parent The parent of the original object to be copied
+         * @returns {Promise} resolved with an array of clones of the models
+         * of the object tree being copied. Copying is done in a bottom-up
+         * fashion, so that the last member in the array is a clone of the model
+         * object being copied. The clones are all full composed with
+         * references to their own children.
+         */
+        CopyService.prototype.buildCopyPlan = function(domainObject, parent, progress) {
```
Comment (VWoeltjen) [inline platform/entanglement/src/services/CopyService.js]: I'd mark this `@private`, don't think it makes sense outside of the broader copying workflow.

File: platform/entanglement/src/services/CopyService.js [hunk 6]
```diff
@@ -54,45 +57,156 @@ define(
...
+         * @param parent
+         * @param progress
+         * @returns a promise that will be completed with the clone of
+         * domainObject when the duplication is successful.
+         */
+        CopyService.prototype.perform = function (domainObject, parent) {
+            var $q = this.$q,
+                deferred = $q.defer();
+            if (this.validate(domainObject, parent)) {
+                this.buildCopyPlan(domainObject, parent, deferred.notify)
```
Comment (VWoeltjen) [inline platform/entanglement/src/services/CopyService.js]: I'd softly recommend creating a `CopyTask` class or similar to encapsulate this process. There are a few new private methods on `CopyService` that all need to pass around a similar set of objects as arguments, which can be a little hard to follow and may be error-prone; think you could move this all into a `new CopyTask(domainObject, parent, deferred).run()`. Internally, `CopyTask` would have `buildCopyPlan`, `persistObjects`, `addClonesToParent`, but they'd all be looking at instance variables instead of arguments, making it a little easier to follow/maintain (instead of looking at each method and its usages to figure out what's what, just go to the constructor)

File: platform/entanglement/src/actions/CopyAction.js [hunk 7]
```diff
@@ -34,16 +34,102 @@ define(
...
+        /**
+         * Executes the CopyAction. The CopyAction uses the default behaviour of 
+         * the AbstractComposeAction, but extends it to support notification 
+         * updates of progress on copy.
+         */
+        CopyAction.prototype.perform = function() {
+            var self = this;
+
+            return AbstractComposeAction.prototype.perform.call(this)
+                .then(
```
Comment (VWoeltjen) [inline platform/entanglement/src/actions/CopyAction.js]: A nit, but to reduce the nesting I'd move these three function declarations out of the `then` call, and just say `then(success, error, notification)`.

**Comment by VWoeltjen** [comment]
Finished review (see line comments), reassigning back to @akhenry for follow up.

Overall this is looking good, and the new copy process seems well thought out. Think this is ready to merge once the smaller and/or functional changes are made (code style, consistency with existing behavior); would be fine leaving some of the larger changes (API changes to `DialogService`, `CopyTask`) and creating follow-up issues for those, if they seem like they'd be time-consuming.

**Comment by VWoeltjen** [comment]
For linking: Addresses #127

File: platform/entanglement/src/actions/CopyAction.js (hunk 2 above)
Comment (akhenry) [inline platform/entanglement/src/actions/CopyAction.js]: Yep, you're right. Nothing has fundamentally changed in the way that the dialogService itself (the dismiss function always worked this way) but now that we have asynchronous processes potentially launching and dismissing dialogs we need better tracking of dialog instances.

I don't remember now why I marked dismiss() private though to be honest, that may have been by mistake.

File: platform/entanglement/src/services/CopyService.js (hunk 3 above)
Comment (akhenry) [inline platform/entanglement/src/services/CopyService.js]: Oh man, nesting gets way too deep here, so I welcome any suggestions for improving readability. I'd like to revisit this function post-#215 actually and see if it can't be refactored to be a little cleaner.

File: platform/entanglement/src/services/CopyService.js (hunk 4 above)
Comment (akhenry) [inline platform/entanglement/src/services/CopyService.js]: Ah ok, this is due to me misinterpreting how linking works I think. In my original implementation B' __was__ located inside C', but this was in effect converting links into first class objects, which I figured was undesirable. I 'fixed' it based on my reading of how links are determined by LocationCapability.isLink(). 

In that case, I'm not quite sure how links should be handled on copy. If I am reading your explanation correctly, B' would be contained in __and located in__ C'. This would have the effect of elevating C' to an original object though?

File: platform/entanglement/src/services/CopyService.js (hunk 6 above)
Comment (akhenry) [inline platform/entanglement/src/services/CopyService.js]: Fair point, I'll refactor it.

File: platform/entanglement/src/services/CopyService.js (hunk 4 above)
Comment (VWoeltjen) [inline platform/entanglement/src/services/CopyService.js]: That is correct, links become originals currently. This is counter-intuitive to me as well, but during design sessions this was what users indicated they would expect. We may end up refining that after feedback (Move/Copy/Link hasn't had much actual user time)

The only case where we'd want to use links instead of creating new originals is if objects are not creatable (in that case, links should be made.) This has not been implemented, though, and I'd consider that out of scope for this issue.

Comment (akhenry) [inline platform/entanglement/src/services/CopyService.js]: Ok, understood. I was trying to be too clever :) Will revert it to the previous behaviour.

File: platform/entanglement/src/actions/CopyAction.js (hunk 2 above)
Comment (akhenry) [inline platform/entanglement/src/actions/CopyAction.js]: Created #254 to track this.

Comment (akhenry) [inline platform/entanglement/src/actions/CopyAction.js]: I take it back, dismiss() is not marked private, but I suspect it is intended to be private as it has no jsdoc and was only rendered 'public' by previous refactoring of the class to prototypal function declarations.

File: platform/entanglement/src/actions/CopyAction.js (hunk 7 above)
Comment (akhenry) [inline platform/entanglement/src/actions/CopyAction.js]: Done.

File: platform/entanglement/src/services/CopyService.js (hunk 6 above)
Comment (akhenry) [inline platform/entanglement/src/services/CopyService.js]: Refactored out to new CopyTask

File: platform/entanglement/src/services/CopyService.js (hunk 5 above)
Comment (akhenry) [inline platform/entanglement/src/services/CopyService.js]: Done

**Comment by akhenry** [comment]
Created https://github.com/nasa/openmctweb/issues/256 to handle refactoring for new object creation changes

**Comment by akhenry** [comment]
Comments addressed, and CI build passing.

**Comment by VWoeltjen** [comment]
### Reviewer Checklist

1. Changes appear to address issue? Y
2. Appropriate unit tests included? Y
3. Code style and in-line documentation are appropriate? Y
4. Commit messages meet standards? Y
